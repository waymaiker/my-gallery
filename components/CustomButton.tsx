import React from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";
import { Icons } from "../constants/constants";

type CustomButtonAttributesType = {
  action: Function,
  text?: string,
  isDisabled?: boolean,
  withIcon?: string
};

export default function CustomButton({action, text, isDisabled, withIcon}: CustomButtonAttributesType) {
  return <TouchableOpacity
    style={ withIcon == undefined ? [isDisabled ? styles.buttonDisabled : styles.buttonClose, styles.button] : {}}
    onPress={ isDisabled ? () => {} : () => action()}
  >
    {
      withIcon == undefined
        ? <Text style={styles.textStyle}> {text} </Text>
        : Icons[withIcon]
    }
  </TouchableOpacity>
}

const styles = StyleSheet.create({
  button: {
    borderRadius: 10,
    padding: 10,
    margin: 5,
    elevation: 2,
  },
  buttonClose: {
    backgroundColor: '#2196F3',
  },
  buttonDisabled: {
    opacity: 0.5,
    backgroundColor: '#2196F3',
  },
  textStyle: {
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});