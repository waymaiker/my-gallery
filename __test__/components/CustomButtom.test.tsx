import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import CustomButton from '../../components/CustomButton';

const onClick = jest.fn();

describe('<CustomButton />', () => {
  it('should match the snapshot', () => {    
    const rendered = render(<CustomButton action={onClick}/>);
    expect(rendered).toMatchSnapshot();
  });

  it('should display label "Click Here"', () => {
    const rendered = render(<CustomButton action={onClick} text="Click Here"/>);
    const textComponent = rendered.getByTestId('custom-button');
    expect(textComponent.props.children).toMatchSnapshot()
  });

  it('should display a Disabled button', () => {
    const rendered = render(<CustomButton action={onClick} isDisabled={true}/>);
    const textComponent = rendered.getByTestId('custom-button');

    fireEvent.press(textComponent)
    
    expect(onClick).toBeCalledTimes(0)
    expect(textComponent.props.style.opacity).toEqual(0.5)
  });

  it('Button should be clickable', () => {
    const rendered = render(<CustomButton action={onClick} isDisabled={false}/>);
    const textComponent = rendered.getByTestId('custom-button');

    fireEvent.press(textComponent)
    
    expect(onClick).toBeCalledTimes(1)
    expect(textComponent.props.style.opacity).toEqual(1)
  });
})