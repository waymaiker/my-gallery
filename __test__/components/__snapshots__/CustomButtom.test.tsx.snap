// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomButton /> should display label "Click Here" 1`] = `
[
  <Text
    style={
      {
        "color": "white",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
     
    Click Here
     
  </Text>,
  <PressabilityDebugView
    color="cyan"
  />,
]
`;

exports[`<CustomButton /> should match the snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#2196F3",
      "borderRadius": 10,
      "elevation": 2,
      "margin": 5,
      "opacity": 1,
      "padding": 10,
    }
  }
  testID="custom-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
     
     
  </Text>
</View>
`;
